linker/src/lib.rs

```diff
@@ -93,31 +93,32 @@ pub fn link<T: FieldElement>(graph: PILGraph<T>) -> Result<PILFile<T>, Vec<Strin
             let to = link.to;
 
             // the lhs is `instr_flag { inputs, outputs }`
+            let mut expressions =
+                Vec::with_capacity(1 + from.params.inputs.len() + from.params.outputs.len());
+            //flag
+            expressions.extend(to.operation.id.iter().cloned().map(Expression::Number));
+            // inputs
+            expressions.extend(
+                from.params
+                    .inputs
+                    .into_iter()
+                    .map(|i| {
+                        assert!(i.ty.is_none() || i.ty == Some("write".to_string()));
+                        (i.name, i.index)
+                    })
+                    .map(|(name, index)| index_access(direct_reference(name), index)),
+            );
+            // outputs
+            expressions.extend(from.params.outputs.into_iter().map(|p| match p.ty {
+                None => index_access(direct_reference(p.name), p.index),
+                // write register as output is mapped as a next ref in the plookup
+                Some(s) if s == "write" => index_access(next_reference(p.name), p.index),
```

Comment:
same here re register's type, I think we shouldn't use strings. There must be something in the type